internal/command/run.go

```diff
@@ -100,9 +76,26 @@ func runWithStreamingOutput(exe string, args []string, env []string, outWriter,
 	return true, nil
 }
 
-func collectOutput(rc io.ReadCloser, w io.Writer) {
+func collectOutput(rcStdout, rcStderr io.ReadCloser, wStdout, wStderr io.Writer) error {
+	var stdoutErr, stderrErr error
+	var wg sync.WaitGroup
+	wg.Add(1)
```

Comment:
Not sure what happened but I think I had a comment here, but now it no longer shows.
Is the go routine just for better performance or does it solve some interleave problem?

Reply:
Without the goroutine, you either collect all stdout output first, or all stderr output first. However, often binaries log to "both" stdout and stderr "at the same time", and the natural output in the terminal for example is to see both interleaved. Without the goroutine, the "time order" is not preserved and you see e.g. all stdout before you see stderr.

Apart from that, command execution seems to be blocked in some cases when stderr is not collected. I saw this when running multi stage builds with buildah in debug mode. The build hang and eventually the Tekton task ran into a timeout. When stderr was collected right away (with the goroutine), the command did not hang anymore. Unfortunately I was not able to reproduce this in a test case quickly, but I'm sure there is a way to demonstrate that.